server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1654,10 +1655,12 @@ public void flush(boolean force, boolean waitIfOngoing) throws EngineException {
             }
             try {
                 // Only flush if (1) Lucene has uncommitted docs, or (2) forced by caller, or (3) the
-                // newly created commit points to a different translog generation (can free translog)
+                // newly created commit points to a different translog generation (can free translog),
+                // or (4) the processed local checkpoint is ahead of the local checkpoint tracked in the safe commit
```

Comment:
It's unclear what the safe commit is in the case of calling this method. Instead I would say something along the lines of:
// or (4) the local checkpoint information in the last commit is stale, which slows down future recoveries.